ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/validation/TransactionConsistencyValidator.scala

```diff
@@ -6,82 +6,75 @@ package com.daml.ledger.participant.state.kvutils.committer.transaction.validati
 import com.daml.ledger.participant.state.kvutils.committer.transaction.{
   DamlTransactionEntrySummary,
   Rejections,
   Step,
 }
 import com.daml.ledger.participant.state.kvutils.committer.{CommitContext, StepContinue, StepResult}
 import com.daml.ledger.participant.state.v1.RejectionReasonV0
-import com.daml.lf.data.Time.Timestamp
 import com.daml.lf.transaction.Transaction.{
   DuplicateKeys,
   InconsistentKeys,
   KeyActive,
   KeyCreate,
   NegativeKeyLookup,
 }
+import com.daml.lf.value.Value
 import com.daml.logging.LoggingContext
 
-private[transaction] object ContractKeysValidator extends TransactionValidator {
+private[transaction] object TransactionConsistencyValidator extends TransactionValidator {
 
-  /** Creates a committer step that validates casual monotonicity and consistency of contract keys
-    * against the current ledger state.
+  /** Validates consistency of contracts and contract keys against the current ledger state.
+    * For contracts, checks whether all contracts used in the transaction are still active.
+    * For keys, checks whether they are consistent and there are no duplicates.
+    *
+    * @param rejections A helper object for creating rejection [[Step]]s.
+    * @return A committer [[Step]] that performs validation.
     */
   override def createValidationStep(rejections: Rejections): Step = new Step {
     def apply(
         commitContext: CommitContext,
         transactionEntry: DamlTransactionEntrySummary,
     )(implicit loggingContext: LoggingContext): StepResult[DamlTransactionEntrySummary] = {
-      val damlState = commitContext
-        .collectInputs[(DamlStateKey, DamlStateValue), Map[DamlStateKey, DamlStateValue]] {
-          case (key, Some(value)) if key.hasContractKey => key -> value
-        }
-      val contractKeyDamlStateKeysToContractIds: Map[DamlStateKey, RawContractId] =
-        damlState.collect {
-          case (k, v) if k.hasContractKey && v.getContractKeyState.getContractId.nonEmpty =>
-            k -> v.getContractKeyState.getContractId
-        }
-      // State before the transaction
-      val contractKeyDamlStateKeys: Set[DamlStateKey] =
-        contractKeyDamlStateKeysToContractIds.keySet
-      val contractKeysToContractIds: Map[DamlContractKey, RawContractId] =
-        contractKeyDamlStateKeysToContractIds.map(m => m._1.getContractKey -> m._2)
-
       for {
-        stateAfterMonotonicityCheck <- checkContractKeysCausalMonotonicity(
-          commitContext.recordTime,
-          contractKeyDamlStateKeys,
-          damlState,
+        stepResult <- validateConsistencyOfKeys(
+          commitContext,
           transactionEntry,
           rejections,
         )
-        finalState <- performTraversalContractKeysChecks(
-          commitContext.recordTime,
-          contractKeysToContractIds,
-          stateAfterMonotonicityCheck,
+        finalStepResult <- validateConsistencyOfContracts(
```

Comment:
did you deliberately check key consistency before contract consistency? I don’t think it matters apart from error messages and I haven’t really thought through which order gives us better error messages so just curious.

Reply:
I was also thinking about this. However, I just picked the order from the description: https://github.com/digital-asset/daml/pull/10066#discussion_r664555881 Shall we revisit this?

Reply:
I think this order makes sense or is at least fine. And in the case where a contract key changed value and the old assignment is no longer active, I think it’s more useful to report this as an InconsistentKeys than an inactive contract and that’s what your implementation does :+1: